File: 2023/fsharp/day04a/Program.fs
open System
open System.IO

let usage () =
    let progname = Path.GetFileName(Environment.GetCommandLineArgs().[0])
    printfn "usage: %s <file>" progname
    Environment.Exit(1)

let processInput (contents: string) =
    Seq.fold
        (fun (result: uint) (line: string) ->
            let rest = line.Split(": ")[1]
            let winningStr = rest.Split(" | ")[0]
            let winningSet =
                Seq.fold
                    (fun set numStr ->
                        let num = Int32.Parse(numStr)
                        Set.add num set)
                    Set.empty
                    (winningStr.Split (" ", StringSplitOptions.RemoveEmptyEntries))
            let handStr = rest.Split(" | ")[1]
            let handSet =
                Seq.fold
                    (fun set numStr ->
                        let num = Int32.Parse(numStr)
                        Set.add num set)
                    Set.empty
                    (handStr.Split (" ", StringSplitOptions.RemoveEmptyEntries))
            let intersection = Set.intersect winningSet handSet
            let count = Set.count intersection
            result + if (count > 0) then (pown 2u (count - 1)) else 0u)
        0u
        (contents.Split ('\n', StringSplitOptions.RemoveEmptyEntries))

[<EntryPoint>]
let main argv =
    if argv.Length < 1 then
        usage()
    let filename = argv.[0]
    let contents = File.ReadAllText(filename)
    let result = processInput contents
    printfn "result = %d" result
    0

File: 2023/csharp/day04a/Program.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Usage()
    {
        string? progPath = Environment.ProcessPath;
        string? progName = Path.GetFileName(progPath);
        Console.WriteLine("usage: " + progName + " <file>");
        System.Environment.Exit(1);
    }

    static uint Process(string contents)
    {
        uint result = 0;
        foreach (string line in contents.Split('\n'))
        {
            string[] lineParts = line.Split(": ");
            string rest = lineParts[1];
            string[] numberParts = rest.Split(" | ");
            string winningStr = numberParts[0];
            HashSet<uint> winningSet = new HashSet<uint>();
            foreach (string numStr in winningStr.Split(' '))
            {
                if (numStr != "")
                {
                    uint num = UInt32.Parse(numStr);
                    winningSet.Add(num);
                }
            }
            string handStr = numberParts[1];
            HashSet<uint> handSet = new HashSet<uint>();
            foreach (string numStr in handStr.Split(' '))
            {
                if (numStr != "")
                {
                    uint num = UInt32.Parse(numStr);
                    handSet.Add(num);
                }
            }
            HashSet<uint> intersection = new HashSet<uint>(winningSet);
            intersection.IntersectWith(handSet);
            int count = intersection.Count;
            if (count > 0u)
            {
                result += (uint)(1 << (count - 1));
            }
        }
        return result;
    }

    public static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Program.Usage();
        }
        string fileName = args[0];
        string contents = File.ReadAllText(fileName);
        contents = contents.Remove(contents.Length - 1, 1);
        uint result = Process(contents);
        Console.WriteLine("result = " + result);
    }
}
